DadMakeAFunny/app/src/main/java/com/example/dadmakeafunny/JokesReader.java

```diff
@@ -0,0 +1,130 @@
+package com.example.dadmakeafunny;
+import android.os.AsyncTask;
+import android.util.Log;
+
+import java.io.InputStream;
+import java.util.ArrayList;
+import android.util.Log;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.ArrayList;
+
+
+public class JokesReader {
+/* class opens connection to reddit and handles all jokes */
+
+    private String url;
+    private ArrayList<Joke> jokes;
+    private String last_joke_id;
+    private static final String BASE_URL = "https://www.reddit.com/r/dadjokes/.json";
+    private static final String LIMIT_FLAG  = "?limit=";
+    private int limit = 20;
+
+    /*constructor*/
+    public JokesReader(){
+        this.jokes = new ArrayList<>();
+        this.last_joke_id = null;
+    }
+
+
+    /*starts the connection to subreddit and pulls in first 20 jokes*/
+
+    public void startConnection(){
+        try {
+            this.url = createUrl();
+            String data = new JokeGetterTask().execute(this.url).get();
+            this.pullJokes(data);
+        }catch (Exception e){
+            Log.e("error",e.toString());
+        }
+    }
+
+    //create a url based on if accessing first time or new access
+    private String createUrl(){
+        if (this.last_joke_id == null){
+            return (this.BASE_URL + this.LIMIT_FLAG + this.limit);
+        }
+        else{
+            return (this.BASE_URL + "?after=" + this.last_joke_id + this.LIMIT_FLAG + this.limit);
+        }
+    }
+
+    //coverts to JSON and read jokes from it
+    private void pullJokes(String data) {
+        try {
+            JSONObject raw_data = new JSONObject(data).getJSONObject("data");
+            JSONArray children = raw_data.getJSONArray("children");
+            this.last_joke_id = raw_data.getString("after");
+            for (int i = 0; i < children.length(); i++) {
+                JSONObject current_json = children.getJSONObject(i).getJSONObject("data");
+
+                Joke curr_joke = new Joke();
+                curr_joke.setTitle(current_json.optString("title"));
+                curr_joke.setLink(current_json.optString("url"));
+                //this might be null
+                curr_joke.setText(current_json.optString("selftext"));
```

Comment:
Once we have that constructor for the `Joke` class, we can use one call to that instead of all of these sets